header-only/awkward/utils.h

```diff
@@ -12,135 +12,108 @@
+      /// is a 32-bit signed integer.
+      return "int32";
+    } else if(std::is_same_v<T, uint32_t>) {
+      /// @brief Returns `uint32` string when the primitive type
+      /// is a 32-bit unsigned integer.
+      return "uint32";
+    } else if(std::is_same_v<T, int64_t>) {
+      /// @brief Returns `int64` string when the primitive type
+      /// is a 64-bit signed integer.
+      return "int64";
+    } else if(std::is_same_v<T, uint64_t>) {
+      /// @brief Returns `uint64` string when the primitive type
+      /// is a 64-bit unsigned integer.
+      return "uint64";
+    } else if(std::is_same_v<T, float>) {
+      /// @brief Returns `float32` string when the primitive type
+      /// is a floating point.
+      return "float32";
+    } else if(std::is_same_v<T, double>) {
+      /// @brief Returns `float32` string when the primitive type
+      /// is a double floating point.
+      return "float64";
+    } else if (std::is_same_v<T, std::complex<float>>) {
+      /// @brief Returns `complex64` string when the primitive type is a
+      /// complex number with float32 real and float32 imaginary parts.
+      return "complex64";
+    } else if (std::is_same_v<T, std::complex<double>>) {
+      /// @brief Returns `complex128` string when the primitive type is a
+      /// complex number with float64 real and float64 imaginary parts.
+      return "complex128";
+    } else if(std::is_same_v<T, short int>) { // at least 16
+      /// @brief Returns `int16` string when the primitive type
+      /// is a 16-bit signed integer.
+      return "int16";
+    } else if(std::is_same_v<T, unsigned short int>) { // at least 16
+      /// @brief Returns `uint16` string when the primitive type
+      /// is a 16-bit unsigned integer.
+      return "uint16";
+    } else if(std::is_same_v<T, int>) { // at least 16
+      /// @brief Returns `int32` string when the primitive type
+      /// is a 32-bit signed integer.
+      return "int32";
+    } else if(std::is_same_v<T, unsigned int>) { // at least 16
+      /// @brief Returns `uint32` string when the primitive type
+      /// is a 32-bit unsigned integer.
+      return "uint32";
+    } else if(std::is_same_v<T, long int>) { // at least 32
+      /// @brief Returns `int64` string when the primitive type
+      /// is a 64-bit signed integer.
+      return "int64";
+    } else if(std::is_same_v<T, unsigned long int>) {
+      /// @brief Returns `uint64` string when the primitive type
+      /// is a 64-bit unsigned integer.
+      return "uint64";
+    } else if(std::is_same_v<T, long long int>) { // at least 64
+      /// @brief Returns `int64` string when the primitive type
+      /// is a 64-bit signed integer.
+      return "int64";
+    } else if(std::is_same_v<T, unsigned long long int>) {
+      /// @brief Returns `uint64` string when the primitive type
+      /// is a 64-bit unsigned integer.
+      return "uint64";
+    } else {
+      return "unsupported_primitive_type";
+    }
   }
```

Comment:
To avoid platform dependence, we have to _not_ specify these in terms of C builtin types. For instance, `signed char` is defined as "at least 8 bits" and its `if` statement comes before the larger types. So if this code runs on a platform in which `signed char` has 16 bits, `type_to_name` would return the string `"int8"` and something would break (possibly without an explicit error message, unless the length is also constrained).

And since `std::is_same_v<T, bool>` checks compatibility/equivalence, rather than exact alias names, testing for `bool` in an `if` statement before `int8_t` would result in any 8-bit data being presented as booleans. That might include ASCII character strings or union tags, which would be very wrong. So for the types that have to be distinguished by exact alias names, I'll reinstate the template-instantiation method.

```suggestion
    if (std::is_same_v<T, int8_t>) {
      return "int8";
    }
    else if (std::is_same_v<T, uint8_t>) {
      return "uint8";
    }
    else if (std::is_same_v<T, int16_t>) {
      return "int16";
    }
    else if (std::is_same_v<T, uint16_t>) {
      return "uint16";
    }
    else if (std::is_same_v<T, int32_t>) {
      return "int32";
    }
    else if (std::is_same_v<T, uint32_t>) {
      return "uint32";
    }
    else if (std::is_same_v<T, int64_t>) {
      return "int64";
    }
    else if (std::is_same_v<T, uint64_t>) {
      return "uint64";
    }
    else if (std::is_same_v<T, float>) {
      return "float32";
    }
    else if (std::is_same_v<T, double>) {
      return "float64";
    }
    else if (std::is_same_v<T, std::complex<float>>) {
      return "complex64";
    }
    else if (std::is_same_v<T, std::complex<double>>) {
      return "complex128";
    }
    else {
      return "unsupported_primitive_type";
    }
  }

  template <>
  const std::string
  type_to_name<bool>() {
    // This takes precedence over the unspecialized template, and therefore any
    // 8-bit data that is not named bool will be mapped to "int8" or "uint8".
    return "bool";
  }
```